controllers/fleetnotification/fleetnotification_controller.go

```diff
@@ -55,21 +63,40 @@ func (r *ManagedFleetNotificationReconciler) Reconcile(ctx context.Context, requ
 	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
 	reqLogger.Info("Reconciling FleetNotification")
 
-	// TODO(user): your logic here
 	nr := ocmagentv1alpha1.ManagedFleetNotificationRecord{}
 
 	err := r.Client.Get(ctx, request.NamespacedName, &nr)
 	if err != nil {
 		return reconcile.Result{}, err
 	}
 
+	for n, rn := range nr.Status.NotificationRecordByName {
+		resendWait := rn.ResendWait
+		for i, ri := range rn.NotificationRecordItems {
+			// Consider the record is stale if the lastSendTime is older than resendWait + 15 days
+			eol := ri.LastTransitionTime.Time.Add(time.Duration(resendWait+NotificationRecordStaleTimeoutInHour) * time.Hour)
+			if time.Now().After(eol) {
+				log.Info(fmt.Sprintf("NotificationRecord for notification %s and hostedcluster %s has not been updated "+
+					"for %d hours and considered as stale, cleaning up...", rn.NotificationName, ri.HostedClusterID,
+					resendWait+NotificationRecordStaleTimeoutInHour))
+
+				patch := []byte(fmt.Sprintf(`[{"op": "remove", "path": "/status/notificationRecordByName/%d/notificationRecordItems/%d"}]`, n, i))
+
+				err = r.Client.Status().Patch(ctx, &nr, client.RawPatch(types.JSONPatchType, patch))
```

Comment:
I think we could add a small test to return failure if the `Patch` call fails? Currently that isn't covered I think. Considering it's one of the important bits that the controller is doing, we could add that test too if possible. Let me know your thoughts.

Reply:
I cannot see much value of adding a failure test, the unit test should be more focusing on the logic, that where the func calls should happen instead of it should fail or succeed.

And I am not sure if the mock client could really do the failure test as well.

Reply:
Thanks Meng Bo. Yup, the test isn't a crucial one. Fine to not have for now.